### DISCUSSION TITLE
How to recalculate the start_angle and end_angle of Arc

### DISCUSSION THREAD

**Discussion by chibai** [mozman/ezdxf#385] [Q&A] [answered]
if Arc.dxf.extrusion != (0,0,1), then we should use ocs._to_wcs() to recalculate Arc.dxf.center
While the Arc.start_point and Arc.end_point are still in wcs
But the Arc.dxf.start_angle and Arc.dxf.end_angle are wrong in wcs

I want to draw the Arc entities by pillow with center, radius, start_angle and end_angle
And I believe the implementation in add-one/drawing is what I want, but I just can't find where it is

Above all, my question is， How can I recalculate the start_agngle and end_angle in wcs with Arc.dxf.extrusion != (0,0,1), Thank you very much

**Comment by mozman** [comment, marked as answer]
This is not a simple question, but I assume if you use `Pillow` the problem is reduced to a 2D only scenario. In 2D there are only two OCS (extrusion vectors) where an ARC projects as an ARC onto the xy-plane:
1. the default case (0, 0, 1) - no angle conversion necessary
2. (0, 0, -1) the reflected default extrusion
3. every other extrusion vector projects the circular arc as an elliptic arc onto the xy-plane

This code shows the conversion of angles from the reflected OCS (0, 0, -1) into the default OCS == WCS

```Python
from ezdxf.math import OCS, linspace, Vec3

source_ocs = OCS((0, 0, -1))
for angle in linspace(0, 360, 37):
    # convert source angle into a WCS vector
    v = source_ocs.to_wcs(Vec3.from_deg_angle(angle))
    # This works only for the extrusion vector (0, 0, -1)
    print(f"Source angle: {angle}; Target angle: {v.angle_deg:f} == {180-angle:f} (180-angle)")

```

The x-axis is 0 degree and the arc goes from start to end in counter clockwise orientation. The simple formula is: `180-angle`.

We avoided this problem in the drawing add-on by converting circles, arcs and ellipses into cubic Bèzier curves (`Path` object), which are often supported by various backends, but not by the `Pillow` package.

This example shows how to convert a DXF entity (CIRCLE, ARC, ELLIPSE, SPLINE) into an ezdxf  `Path` object and approximate this curve by a polyline. This works for every extrusion vector and therefore in 2D and 3D:

```Python
import ezdxf
from ezdxf.path import make_path

doc = ezdxf.new()
msp = doc.modelspace()
arc = msp.add_arc((0, 0), radius=1, start_angle=0, end_angle=90)

# create a Path() object from the ARC
p = make_path(arc)
vertices = [
    # projection onto the xy-plan by ignoring the z-axis
    (vertex.x, vertex.y) for vertex in p.flattening(distance=0.01)
]  # adjust "distance" to your precision needs, max. distance between ARC and chord (approximation line)

msp.add_lwpolyline(vertices)  # Pillow: draw.line(vertices)
```
The ARC approximated by a LWPOLYLINE:
![image](https://user-images.githubusercontent.com/4230966/113423160-4085f900-93ce-11eb-89a1-3ffb09c1f9bf.png)

**Reply by chibai** [reply]
Thanks for your compelet answer~
I'm in he Tomb-sweeping Day(Qingming Holiday), that's why I reply late
